File: 43. Multiply Strings/43. Multiply Strings.cpp
class Solution {
public:
    string multiply(string num1, string num2) {
        int n1 = num1.size(), n2 = num2.size();
        string res(n1 + n2, '0');
        
        for (int i = n1 - 1; i >= 0; i--) {
            for (int j = n2 - 1; j >= 0; j--) {
                int mul = (num1[i] - '0') * (num2[j] - '0');
                int sum = (res[i + j + 1] - '0') + mul;
                
                res[i + j + 1] = sum % 10 + '0';
                res[i + j] += sum / 10;
            }
        }
        
        size_t startpos = res.find_first_not_of("0");
        if (string::npos != startpos) {
            return res.substr(startpos);
        }
        
        return "0";
    }
};

File: 43. Multiply Strings/43. Multiply Strings.rs
impl Solution {
    pub fn multiply(num1: String, num2: String) -> String {
        let n1 = num1.len();
        let n2 = num2.len();
        let mut res = vec![0; n1 + n2];
        
        for (i, c1) in num1.chars().rev().enumerate() {
            for (j, c2) in num2.chars().rev().enumerate() {
                let mul = (c1 as u8 - b'0') * (c2 as u8 - b'0');
                let sum = res[i + j] + mul as i32;
                
                res[i + j] = sum % 10;
                res[i + j + 1] += sum / 10;
            }
        }
        
        while res.len() > 1 && res.last() == Some(&0) {
            res.pop();
        }
        
        res.into_iter().rev().map(|d| (d as u8 + b'0') as char).collect()
    }
}